File: 2024/cpp/day11.cpp
#include <cstdio>
#include <fstream>
#include <iostream>
#include <sstream>
#include <string>
#include <sys/types.h>
#include <unordered_map>
#include <vector>

using namespace std;

long int part1();
long int part2();

int main() {
  cout << part1() << endl;
  cout << part2() << endl;
}

long int part1() {
  string l;
  ifstream input("../data/input11.txt");
  vector<int> line;

  if (input.is_open()) {
    getline(input, l);
    stringstream ss(l);
    string each;
    while (ss >> each) {
      line.push_back(stoi(each));
    }
    input.close();
  }

  int blinks = 25;
  unordered_map<long int, long int> counts;

  for (const auto &d : line) {
    counts[d]++;
  }

  for (int i = 0; i < blinks; i++) {
    unordered_map<long int, long int> new_counts;
    new_counts.reserve(counts.size() * 2);
    new_counts[1] = counts[0];

    for (const auto &entry : counts) {
      long int num = entry.first, count = entry.second;

      if (num != 0) {
        string s = to_string(num);
        const size_t size = s.size();

        if (size % 2 == 1) {
          new_counts[2024 * num] += count;
        } else {
          const size_t mid = size / 2;
          new_counts[stol(s.substr(0, mid))] += count;
          new_counts[stol(s.substr(mid))] += count;
        }
      }
    }

    counts = std::move(new_counts);
  }

  long int sum = 0;
  for (const auto &keyval : counts) {
    sum += keyval.second;
  }

  return sum;
}

long int part2() {
  string l;
  ifstream input("../data/input11.txt");
  vector<int> line;

  if (input.is_open()) {
    getline(input, l);
    stringstream ss(l);
    string each;
    while (ss >> each) {
      line.push_back(stoi(each));
    }
    input.close();
  }

  int blinks = 75;
  unordered_map<long int, long int> counts;

  for (const auto &d : line) {
    counts[d]++;
  }

  for (int i = 0; i < blinks; i++) {
    unordered_map<long int, long int> new_counts;
    new_counts.reserve(counts.size() * 2);
    new_counts[1] = counts[0];

    for (const auto &entry : counts) {
      long int num = entry.first, count = entry.second;

      if (num != 0) {
        string s = to_string(num);
        const size_t size = s.size();

        if (size % 2 == 1) {
          new_counts[2024 * num] += count;
        } else {
          const size_t mid = size / 2;
          new_counts[stol(s.substr(0, mid))] += count;
          new_counts[stol(s.substr(mid))] += count;
        }
      }
    }

    counts = std::move(new_counts);
  }

  long int sum = 0;
  for (const auto &keyval : counts) {
    sum += keyval.second;
  }

  return sum;
}

File: 2024/python/day11.py
from functools import cache


@cache
def process_number(num: int) -> tuple[int, ...]:
    if num == 0:
        return (1,)
    s = str(num)
    if len(s) % 2 == 1:
        return (2024 * num,)
    mid = len(s) // 2
    left, right = int(s[:mid]), int(s[mid:])
    return (left, right)


@cache
def simulate(line: tuple[int], blinks: int) -> int:
    counts = dict.fromkeys(line, 0)
    for v in line:
        counts[v] += 1

    for _ in range(blinks):
        new_counts = {}
        for num, count in counts.items():
            for new_num in process_number(num):
                new_counts[new_num] = new_counts.get(new_num, 0) + count
        counts = new_counts

    return sum(counts.values())


def part1() -> int:
    with open("../data/input11.txt", "r") as inp:
        line = tuple(map(int, inp.read().strip().split()))

    return simulate(line, 25)


def part2() -> int:
    with open("../data/input11.txt", "r") as inp:
        line = tuple(map(int, inp.read().strip().split()))

    return simulate(line, 75)


print(part1())
print(part2())
